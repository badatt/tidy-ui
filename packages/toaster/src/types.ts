import React from 'react';
import { IBaseProps } from '@tidy-ui/types';

/**
 * Closable
 */
interface IClosable {
  /**
   * Callback (if any) after the message widget is closed
   *
   * @type {?((e?: any | undefined) => void)}
   */
  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
  onClose?: (e?: any | undefined) => void;
}

/**
 * ToastItem
 */
export type TToastItem = React.ReactElement<IClosable>;

/**
 * Toast
 */
export interface IToast {
  /**
   * Id
   */
  id?: string;
  /**
   * Item
   */
  item?: TToastItem;
}

/**
 * Toaster props
 */
export interface IToasterProps extends IBaseProps {
  /**
   * Maximum number of toasts that a toaster can toss at any instant
   */
  limit?: number;
  /**
   * Timeout for toasts, in milliseconds
   *
   * @default 5000
   */
  timeout?: number | 'infinite';
}
