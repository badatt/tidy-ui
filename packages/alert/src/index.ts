export * from './Alert';
export * from './AlertFooter';
export * from './AlertTitle';
