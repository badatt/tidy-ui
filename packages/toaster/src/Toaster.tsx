import React from 'react';
import { styled, usePortal } from '@tidy-ui/commons';
import { IToast, IToasterProps } from './types';

const ToasterRoot = styled.div`
  position: absolute;
  display: block;
  bottom: 8px;
`;

const ToastItem = styled.div`
  margin: 0.5rem 1rem;
  transition: all cubic-bezier(0.075, 0.82, 0.165, 1);
`;

/**
 * Toaster
 *
 * @param {IToasterProps} props toaster props
 * @returns {React.FC} react component
 */
const Toaster: React.FC<IToasterProps> = (props: IToasterProps) => {
  const { timeout } = props;
  const [toasts, setToasts] = React.useState<IToast[]>([]);
  const { Portal } = usePortal();

  return (
    <Portal>
      <ToasterRoot role="toaster">
        {toasts?.map((v, i) => {
          return <ToastItem key={i}>{v.item}</ToastItem>;
        })}
      </ToasterRoot>
    </Portal>
  );
};

Toaster.defaultProps = {
  timeout: 5000,
};

export { Toaster };
