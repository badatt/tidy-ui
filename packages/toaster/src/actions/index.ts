import { ActionMap } from '@tidy-ui/commons';
import { TToastItem } from '../types';

/**
 * Toaster actions
 */
export enum ToasterActions {
  AddToast = 'ADD_TOAST',
  Clear = 'CLEAR',
  RemoveToast = 'REMOVE_TOAST',
}

/**
 * Toaster actions payloads
 */
type ToasterPayLoad = {
  /**
   * Add toast
   */
  [ToasterActions.AddToast]: {
    /**
     * Item to add
     */
    item?: TToastItem;
  };
  /**
   * Clears all toasts
   */
  [ToasterActions.Clear]: null;
  /**
   * Removes toast
   */
  [ToasterActions.RemoveToast]: {
    /**
     * Toast id
     */
    id?: string;
  };
};

/**
 * Toaster action types
 */
export type ToasterActionTypes = ActionMap<ToasterPayLoad>[keyof ActionMap<ToasterPayLoad>];
