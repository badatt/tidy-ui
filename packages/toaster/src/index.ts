export * from './Toaster';
export { default as useToaster } from './useToaster';
