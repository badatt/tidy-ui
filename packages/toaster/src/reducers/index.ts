import { Queue } from '@tidy-ui/commons';
import { ToasterActions, ToasterActionTypes } from '../actions';
import { IToast } from '../types';
import { IToasterState } from './types';

const queue = new Queue<IToast>();
const toastSize = {
  current: 0,
  /**
   * Decrements toast size
   */
  decrement() {
    this.current -= 1;
  },
  /**
   * Increments toast size
   */
  increment() {
    this.current += 1;
  },
  /**
   * Resets size
   */
  reset() {
    this.current = 0;
  },
};

/**
 * Initial state
 */
const initialState: IToasterState = {
  toasts: [],
};

/**
 * State initializer
 *
 * @param {IToasterState} initState init state
 * @returns {IToasterState} initial state
 */
const initializer = (initState: IToasterState): IToasterState => {
  return { ...initState };
};

/**
 * Toaster reducer
 *
 * @param {IToasterState} prevState previous toaster state
 * @param {ToasterActionTypes} action toaster actions
 * @returns {IToasterState} mutated toaster state
 */
const reducer = (prevState: IToasterState, action: ToasterActionTypes): IToasterState => {
  switch (action.type) {
    case ToasterActions.AddToast: {
      const toasts = prevState.toasts;
      toasts?.push({ id: Date.now().toString(), item: action.payload?.item });
      toastSize.increment();
      return { ...prevState, toasts };
    }
    case ToasterActions.Clear: {
      toastSize.reset();
      return { ...prevState, toasts: [] };
    }
    case ToasterActions.QueueToast: {
      queue.enqueue({
        id: Date.now.toString(),
        item: action.payload?.item,
      });
      return prevState;
    }
    case ToasterActions.RemoveToast: {
      const isExists = prevState.toasts?.some((t) => t.id === action.payload?.id);
      if (isExists) {
        const toasts = prevState.toasts?.filter((t) => t.id !== action.payload?.id);
        toastSize.decrement();
        return { ...prevState, toasts };
      }
      return prevState;
    }
    default:
      return prevState;
  }
};

export { initializer, initialState, queue, reducer, toastSize };
