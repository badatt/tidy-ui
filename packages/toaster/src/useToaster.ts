import * as React from 'react';
import ToasterContext from './provider/ToasterContext';
import { ToasterActions } from './actions';
import { TToastItem } from './types';

/**
 * Toaster
 */
export interface IToaster {
  /**
   * Clears all toasts
   */
  clear: () => void;

  /**
   * Pushes toast
   */
  push: (item: TToastItem) => void;
}

/**
 * useToaster return
 */
export interface IUseToasterFnReturn {
  /**
   * Toaster
   */
  toaster: IToaster;
}

/**
 * useToaster
 *
 * @returns {IUseToasterFnReturn} useToaster return
 */
const useToaster = (): IUseToasterFnReturn => {
  const { dispatch } = React.useContext(ToasterContext);

  const push = React.useCallback(
    (item: TToastItem) => dispatch({ payload: { item }, type: ToasterActions.QueueToast }),
    [],
  );

  const clear = React.useCallback(() => dispatch({ type: ToasterActions.Clear }), []);

  const toaster: IToaster = {
    clear,
    push,
  };

  return { toaster };
};

export default useToaster;
