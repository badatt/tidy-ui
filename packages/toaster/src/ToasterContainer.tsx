import React from 'react';
import { styled } from '@tidy-ui/commons';
import ToasterContext from './provider/ToasterContext';
import { IToasterProps } from './types';
import useToaster from './useToaster';

const ToasterRoot = styled.div`
  position: absolute;
  display: block;
  bottom: 8px;
`;

const ToastItem = styled.div`
  margin: 0.5rem 1rem;
  transition: all cubic-bezier(0.075, 0.82, 0.165, 1);
`;

/**
 * Toaster
 *
 * @param {IToasterProps} props toaster props
 * @returns {JSX.Element} react component
 */
const ToasterContainer: React.FC<IToasterProps> = (props: IToasterProps): JSX.Element => {
  const { timeout } = props;
  const { state } = React.useContext(ToasterContext);
  const { toaster } = useToaster();

  /**
   * Auto remove toast callback
   *
   * @param {string} id to remove toast after timeout
   */
  const autoRemoveToast = (id?: string) => {
    if (typeof timeout === 'number') {
      setTimeout(() => {
        toaster.remove(id);
      }, timeout);
    }
  };

  return (
    <ToasterRoot role="toaster">
      {state.toasts?.map((v, i) => {
        autoRemoveToast(v.id);
        return <ToastItem key={i}>{v.item}</ToastItem>;
      })}
    </ToasterRoot>
  );
};

ToasterContainer.defaultProps = {
  timeout: 5000,
};

export { ToasterContainer };
