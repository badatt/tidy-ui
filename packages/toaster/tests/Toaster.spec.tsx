/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { orchidDark, orchidLight } from '@tidy-ui/commons';
import { Toaster } from '../src';

describe('Toaster', () => {
  it('Basic render', () => {
    const tree = render(
      <ThemeProvider theme={orchidLight}>
        <Toaster>basic</Toaster>
      </ThemeProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('Dark mode basic render', () => {
    const tree = render(
      <ThemeProvider theme={orchidDark}>
        <Toaster>basic</Toaster>
      </ThemeProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
