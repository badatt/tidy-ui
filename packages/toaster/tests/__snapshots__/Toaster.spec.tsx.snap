// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster Basic render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<body>
    <div />
    <div
      class="c0"
      role="toaster"
    />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Toaster Clear all toasts 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<body>
    <div>
      <div>
        <button>
          Add
        </button>
        <button>
          Clear
        </button>
      </div>
    </div>
    <div
      class="c0"
      role="toaster"
    />
  </body>,
  "container": <div>
    <div>
      <button>
        Add
      </button>
      <button>
        Clear
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Toaster Dark mode basic render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<body>
    <div />
    <div
      class="c0"
      role="toaster"
    />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Toaster Push toast 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<body>
    <div>
      <div>
        <button>
          Add
        </button>
        <button>
          Clear
        </button>
      </div>
    </div>
    <div
      class="c0"
      role="toaster"
    />
  </body>,
  "container": <div>
    <div>
      <button>
        Add
      </button>
      <button>
        Clear
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Toaster Toast items in queue 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<body>
    <div>
      <div
        class="c0"
        role="toaster"
      />
    </div>
  </body>,
  "container": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

<div>
    <div
      class="c0"
      role="toaster"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Toaster Toaster with some toasts 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

.c1 {
  margin: 0.5rem 1rem;
}

<body>
    <div>
      <div
        class="c0"
        role="toaster"
      >
        <div
          class="c1"
        >
          <div>
            hello
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": .c0 {
  position: absolute;
  display: block;
  bottom: 8px;
}

.c1 {
  margin: 0.5rem 1rem;
}

<div>
    <div
      class="c0"
      role="toaster"
    >
      <div
        class="c1"
      >
        <div>
          hello
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
